src/backend/xdg_desktop_portal.rs

```diff
@@ -0,0 +1,284 @@
+use std::path::PathBuf;
+
+use crate::backend::DialogFutureType;
+use crate::file_dialog::Filter;
+use crate::{FileDialog, FileHandle};
+
+use ashpd::desktop::file_chooser::{
+    FileChooserProxy, FileFilter, OpenFileOptions, SaveFileOptions,
+};
+// TODO: convert raw_window_handle::RawWindowHandle to ashpd::WindowIdentifier
+use ashpd::{zbus, WindowIdentifier};
+
+use log::warn;
+use smol::block_on;
+
+//
+// Utility functions
+//
+
+fn add_filters_to_open_file_options(
+    filters: Vec<Filter>,
+    mut options: OpenFileOptions,
+) -> OpenFileOptions {
+    for filter in &filters {
+        let mut ashpd_filter = FileFilter::new(&filter.name);
+        for file_extension in &filter.extensions {
+            ashpd_filter = ashpd_filter.glob(&format!("*.{}", file_extension));
+        }
+        options = options.add_filter(ashpd_filter);
+    }
+    options
+}
+
+fn add_filters_to_save_file_options(
+    filters: Vec<Filter>,
+    mut options: SaveFileOptions,
+) -> SaveFileOptions {
+    for filter in &filters {
+        let mut ashpd_filter = FileFilter::new(&filter.name);
+        for file_extension in &filter.extensions {
+            ashpd_filter = ashpd_filter.glob(&format!("*.{}", file_extension));
+        }
+        options = options.add_filter(ashpd_filter);
+    }
+    options
+}
+
+// refer to https://github.com/flatpak/xdg-desktop-portal/issues/213
+fn uri_to_pathbuf(uri: &str) -> Option<PathBuf> {
+    uri.strip_prefix("file://").map(PathBuf::from)
+}
+
+fn unwrap_or_warn<T, E: std::fmt::Debug>(result: Result<T, E>) -> Option<T> {
```

Comment:
I'm not thrilled about the name of this function... any suggestions for alternatives?

Reply:
Yeah, it's not really unwrapping, so I would call it `ok_or_warn`, but I'm not super thrilled about that either.